iceoryx_posh/include/iceoryx_posh/internal/popo/modern_api/base_subscriber.inl

```diff
@@ -104,46 +107,69 @@ inline cxx::expected<cxx::optional<Sample<const T>>, ChunkReceiveError> BaseSubs
-inline void BaseSubscriber<T, port_t>::releaseQueuedSamples() noexcept
+template <typename T, typename Subscriber, typename port_t>
+inline void BaseSubscriber<T, Subscriber, port_t>::releaseQueuedSamples() noexcept
 {
     m_port.releaseQueuedChunks();
 }
 
-template <typename T, typename port_t>
-inline void
-BaseSubscriber<T, port_t>::setConditionVariable(ConditionVariableData* const conditionVariableDataPtr) noexcept
+template <typename T, typename Subscriber, typename port_t>
+inline void BaseSubscriber<T, Subscriber, port_t>::invalidateTrigger(const uint64_t uniqueTriggerId) noexcept
 {
-    m_port.setConditionVariable(conditionVariableDataPtr);
+    if (m_trigger.getUniqueId() == uniqueTriggerId)
+    {
+        m_port.unsetConditionVariable();
+        m_trigger.invalidate();
+    }
 }
 
-template <typename T, typename port_t>
-inline void BaseSubscriber<T, port_t>::unsetConditionVariable() noexcept
+// ============================== Sample Deleter ============================== //
+
+template <typename T, typename Subscriber, typename port_t>
+inline BaseSubscriber<T, Subscriber, port_t>::SubscriberSampleDeleter::SubscriberSampleDeleter(port_t& port)
+    : m_port(std::ref(port))
 {
-    m_port.unsetConditionVariable();
 }
 
-template <typename T, typename port_t>
-inline bool BaseSubscriber<T, port_t>::hasTriggered() const noexcept
+template <typename T, typename Subscriber, typename port_t>
+inline void BaseSubscriber<T, Subscriber, port_t>::SubscriberSampleDeleter::operator()(T* const ptr) const
 {
-    return m_port.hasNewChunks();
+    auto header = mepoo::convertPayloadPointerToChunkHeader(reinterpret_cast<void*>(ptr));
+    m_port.get().releaseChunk(header);
 }
 
-// ============================== Sample Deleter ============================== //
-
-template <typename T, typename port_t>
-inline BaseSubscriber<T, port_t>::SubscriberSampleDeleter::SubscriberSampleDeleter(port_t& port)
-    : m_port(std::ref(port))
+template <typename T, typename Subscriber, typename port_t>
+inline cxx::expected<WaitSetError>
+BaseSubscriber<T, Subscriber, port_t>::attachTo(WaitSet& waitset,
+                                                [[gnu::unused]] const SubscriberEvent subscriberEvent,
+                                                const uint64_t triggerId,
+                                                const Trigger::Callback<Subscriber> callback) noexcept
 {
+    using SelfType = BaseSubscriber<T, Subscriber, port_t>;
+    Subscriber* self = reinterpret_cast<Subscriber*>(this);
+
+    return waitset
+        .acquireTrigger(self,
+                        {this, &BaseSubscriber<T, Subscriber, port_t>::hasNewSamples},
+                        {this, &SelfType::invalidateTrigger},
+                        triggerId,
+                        callback)
+        .and_then([this](TriggerHandle& trigger) {
+            m_trigger = std::move(trigger);
+            m_port.setConditionVariable(m_trigger.getConditionVariableData());
```

Comment:
I am not sure why the port needs to know the conditionVariable. I would have tried to confine it to the TriggerHandle. The previous design might have been the reason...

Reply:
The existing design is the reason.